spec/unit/rust-crypto/OutgoingRequestProcessor.spec.ts

```diff
@@ -0,0 +1,166 @@
+import { Mocked } from "jest-mock";
+import * as RustSdkCryptoJs from "@matrix-org/matrix-sdk-crypto-js";
+import {
+    KeysBackupRequest,
+    KeysClaimRequest,
+    KeysQueryRequest,
+    KeysUploadRequest,
+    RoomMessageRequest,
+    SignatureUploadRequest,
+    ToDeviceRequest,
+} from "@matrix-org/matrix-sdk-crypto-js";
+
+import { TypedEventEmitter } from "../../../src/models/typed-event-emitter";
+import { HttpApiEvent, HttpApiEventHandlerMap, MatrixHttpApi } from "../../../src";
+import { OutgoingRequestProcessor } from "../../../src/rust-crypto/OutgoingRequestProcessor";
+
+describe("OutgoingRequestProcessor", () => {
+    /** the OutgoingRequestProcessor implementation under test */
+    let processor: OutgoingRequestProcessor;
+
+    /** A mock http backend which processor is connected to */
+    let httpBackend: MockHttpBackend;
+
+    /** a mocked-up OlmMachine which processor is connected to */
+    let olmMachine: Mocked<RustSdkCryptoJs.OlmMachine>;
+
+    /** wait for a call to olmMachine.markRequestAsSent */
+    function awaitCallToMarkAsSent(): Promise<void> {
+        return new Promise((resolve, _reject) => {
+            olmMachine.markRequestAsSent.mockImplementationOnce(async () => {
+                console.log("received call to markAsSent");
+                resolve(undefined);
+            });
+        });
+    }
+
+    beforeEach(async () => {
+        httpBackend = new MockHttpBackend();
+
+        const dummyEventEmitter = new TypedEventEmitter<HttpApiEvent, HttpApiEventHandlerMap>();
+        const httpApi = new MatrixHttpApi(dummyEventEmitter, {
+            baseUrl: "https://example.com",
+            prefix: "/_matrix",
+            fetchFn: httpBackend.fetchFn as typeof global.fetch,
+            onlyData: true,
+        });
+
+        olmMachine = {
+            markRequestAsSent: jest.fn(),
+        } as unknown as Mocked<RustSdkCryptoJs.OlmMachine>;
+
+        processor = new OutgoingRequestProcessor(olmMachine, httpApi);
+    });
+
+    /* simple requests that map directly to the request body */
+    const tests: Array<[any, "POST" | "PUT", string]> = [
+        [KeysUploadRequest, "POST", "https://example.com/_matrix/client/v3/keys/upload"],
+        [KeysQueryRequest, "POST", "https://example.com/_matrix/client/v3/keys/query"],
+        [KeysClaimRequest, "POST", "https://example.com/_matrix/client/v3/keys/claim"],
+        [SignatureUploadRequest, "POST", "https://example.com/_matrix/client/v3/keys/signatures/upload"],
+        [KeysBackupRequest, "PUT", "https://example.com/_matrix/client/v3/room_keys/keys"],
+    ];
+
+    for (const [RequestClass, expectedMethod, expectedPath] of tests) {
+        it(`should handle ${RequestClass.name}s`, async () => {
+            const testBody = '{ "foo": "bar" }';
+            const outgoingRequest = new RequestClass("1234", testBody);
+
+            const reqProm = processor.makeOutgoingRequest(outgoingRequest);
```

Comment:
I'd like a comment along the lines of "This is what we are testing" here - it took me a while to figure out what this test is really doing.

Reply:
Yes, sorry about the lack of comments. Have added some. Is it clearer now?

Reply:
Yes, thank you.